### PR TITLE
add ms5837 driver

### PR DESCRIPTION
**Pull request by xn365**
Please use [PX4 Discuss](http://discuss.px4.io/) or [Slack](http://slack.px4.io/) to align on pull requests if necessary. You can then open draft pull requests to get early feedback.

**Describe problem solved by this pull request**
A clear and concise description of the problem this proposed change will solve.
E.g. For this use case I ran into...

**Describe your solution**
A clear and concise description of what you have implemented.

**Describe possible alternatives**
A clear and concise description of alternative solutions or features you've considered.

**Test data / coverage**
How was it tested? What cases were covered? Logs uploaded to https://review.px4.io/ and screenshots of the important plot parts.

**Additional context**
Add any other related context or media.

### CODE REVIEW

**Comment by xn365** [comment]
Because of the hand slip, the following contents have been deleted：
in file "src/drivers/drv_sensor.h " 
#define DRV_IMU_DEVTYPE_ICM42670P 0x2A
PLZ check.

File: src/drivers/drv_sensor.h [hunk 1]
```diff
@@ -78,7 +78,6 @@
 #define DRV_IMU_DEVTYPE_ICM40609D 0x27
 #define DRV_IMU_DEVTYPE_ICM20948 0x28
 #define DRV_IMU_DEVTYPE_ICM42605 0x29
-#define DRV_IMU_DEVTYPE_ICM42670P 0x2A
```
Comment (dagar) [inline src/drivers/drv_sensor.h:81]: restore this line

File: src/drivers/drv_sensor.h [hunk 2]
```diff
@@ -109,6 +108,8 @@
 #define DRV_DIFF_PRESS_DEVTYPE_SDP33    0x4B
 #define DRV_BARO_DEVTYPE_LPS33HW        0x4C
 #define DRV_BARO_DEVTYPE_TCBP001TA      0x4D
+//Tia ADD 5837
```
Comment (dagar) [inline src/drivers/drv_sensor.h:111]: Remove this comment?

File: src/drivers/barometer/ms5837/ms5837_i2c.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+ * @file ms5837_i2c.cpp
+ *
+ * I2C interface for MS5837
+ */
+
+#include <drivers/device/i2c.h>
+
+#include "ms5837.h"
+
+class MS5837_I2C : public device::I2C
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837_i2c.cpp:44]:
```suggestion
class MS5837_I2C : public device::I2C
```

The MS5837_I2C class isn't really needed as the ms5837 is i2c only. You can collapse the entire thing into an MS5837 class. 

For example see some of the i2c magnetometer drivers. https://github.com/PX4/PX4-Autopilot/blob/master/src/drivers/magnetometer/isentek/ist8310/IST8310.hpp

File: src/drivers/barometer/ms5837/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,49 @@
+############################################################################
+#
+#   Copyright (c) 2015-2019 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/barometer/ms5837/CMakeLists.txt:3]:
```suggestion
#   Copyright (c) 2021 PX4 Development Team. All rights reserved.
```

File: src/drivers/drv_sensor.h (hunk 2 above)
Comment (xn365) [reply src/drivers/drv_sensor.h:111]: OK

**Comment by xn365** [comment]
Modify as recommended by Dagar.

File: src/drivers/barometer/ms5837/ms5837.hpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2019 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837.hpp:3]:
```suggestion
 *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
```

File: src/drivers/barometer/ms5837/ms5837.cpp [hunk 6]
```diff
@@ -0,0 +1,451 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2019 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837.cpp:3]:
```suggestion
 *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
```

File: src/drivers/barometer/ms5837/ms5837_main.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2019, 2021 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837_main.cpp:3]:
```suggestion
 *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
```

File: src/drivers/barometer/ms5837/ms5837_registers.h [hunk 8]
```diff
@@ -0,0 +1,115 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2012-2019 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837_registers.h:3]:
```suggestion
 *   Copyright (C) 2021 PX4 Development Team. All rights reserved.
```

File: src/drivers/barometer/ms5837/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	MODULE drivers__barometer__ms5837
+	MAIN ms5837
+	COMPILE_FLAGS
+		-Wno-cast-align # TODO: fix and enable
+	SRCS
+		ms5837.cpp
+		ms5837_main.cpp
+		ms5837.hpp
+	DEPENDS
+		cdev
```
Comment (dagar) [inline src/drivers/barometer/ms5837/CMakeLists.txt:44]:
```suggestion
```

File: src/drivers/barometer/ms5837/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+		drivers__device
```
Comment (dagar) [inline src/drivers/barometer/ms5837/CMakeLists.txt:45]:
```suggestion
```

File: src/drivers/barometer/ms5837/ms5837.hpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	void			RunImpl();
+	void 			print_status() override;
+	int			read(unsigned offset, void *data, unsigned count) override;
+
+protected:
+	int			probe() override;
+
+	PX4Barometer		_px4_barometer;
+
+	device::Device		*_interface;
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837.hpp:84]:
```suggestion
```

Unused now?

File: src/drivers/barometer/ms5837/ms5837.hpp [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	 * When the interval between measurements is greater than the minimum
+	 * measurement interval, a gap is inserted between collection
+	 * and measurement to provide the most recent measurement possible
+	 * at the next interval.
+	 */
+	void			RunImpl();
+	void 			print_status() override;
+	int			read(unsigned offset, void *data, unsigned count) override;
+
+protected:
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837.hpp:79]:
```suggestion
private:
```

File: src/drivers/barometer/ms5837/ms5837.hpp [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+	int			measure();
+
+	/**
+	 * Collect the result of the most recent measurement.
+	 */
+	int			collect();
+
+
+
+
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837.hpp:121]:
```suggestion
```

File: src/drivers/barometer/ms5837/ms5837.hpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+	 */
+	int			measure();
+
+	/**
+	 * Collect the result of the most recent measurement.
+	 */
+	int			collect();
+
+
+
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837.hpp:120]:
```suggestion
```

File: src/drivers/barometer/ms5837/ms5837.hpp [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+	 * @return		OK if the measurement command was successful.
+	 */
+	int			measure();
+
+	/**
+	 * Collect the result of the most recent measurement.
+	 */
+	int			collect();
+
+
```
Comment (dagar) [inline src/drivers/barometer/ms5837/ms5837.hpp:119]:
```suggestion
```

**Comment by dagar** [comment]
There are some minor formatting errors causing CI to fail, but this otherwise looks good to go from my perspective.
https://github.com/PX4/PX4-Autopilot/pull/18213/checks?check_run_id=3568147845
Screenshot from 2021-09-11 11-31-34

Has this driver actually been tested underwater?

**Comment by xn365** [comment]
> Has this driver actually been tested underwater?

Yes.
Ms5837 can be used in both air and water.

**Comment by xn365** [comment]
This drive only gives the absolute pressure reading of the sensor. If it is used in liquid, it is also necessary to calculate the water depth according to the initial conditions (liquid density, air pressure on the liquid surface, etc.).

**Comment by xn365** [comment]
some test:
[https://github.com/xn365/pictures/blob/main/ms5837/command.png](url)
[https://github.com/xn365/pictures/blob/main/ms5837/data.png](url)

**Pull request merged**